StyleCop.Analyzers/StyleCop.Analyzers/OrderingRules/SA1210UsingDirectivesMustBeOrderedAlphabeticallyByNamespace.cs

```diff
@@ -50,108 +54,52 @@ public override ImmutableArray<DiagnosticDescriptor> SupportedDiagnostics
-            {
-                NamespaceDeclarationSyntax namespaceDeclaration = syntax.Parent as NamespaceDeclarationSyntax;
-                usingDirectives = namespaceDeclaration?.Usings;
-            }
-
-            if (!usingDirectives.HasValue)
-            {
-                return;
-            }
-
-            foreach (var usingDirective in usingDirectives)
-            {
-                // we are only interested in nodes before the current node
-                if (usingDirective == syntax)
-                {
-                    break;
-                }
-
-                // ignore using alias directives, since they are handled by SA1209
-                if (usingDirective.Alias != null)
-                {
-                    continue;
-                }
+            var compilationUnit = (CompilationUnitSyntax)context.Node;
 
-                INamespaceSymbol precedingNamespaceSymbol;
-                string precedingTopLevelNamespace = GetTopLevelNamespace(semanticModel, usingDirective, out precedingNamespaceSymbol, context.CancellationToken);
-                if (precedingTopLevelNamespace == null || precedingNamespaceSymbol == null)
-                {
-                    continue;
-                }
+            ProcessUsings(context, compilationUnit.Usings);
+        }
 
-                // compare System namespaces to each other, and non-System namespaces to each other
-                if ("System".Equals(precedingTopLevelNamespace, StringComparison.Ordinal) != systemNamespace)
-                {
-                    continue;
-                }
+        private static void HandleNamespaceDeclaration(SyntaxNodeAnalysisContext context)
+        {
+            var namespaceDeclaration = (NamespaceDeclarationSyntax)context.Node;
 
-                string precedingFullyQualifiedNamespace = precedingNamespaceSymbol.ToDisplayString(SymbolDisplayFormat.FullyQualifiedFormat);
-                if (string.Compare(fullyQualifiedNamespace, precedingFullyQualifiedNamespace, StringComparison.OrdinalIgnoreCase) >= 0)
-                {
-                    continue;
-                }
+            ProcessUsings(context, namespaceDeclaration.Usings);
+        }
 
-                string @namespace = namespaceSymbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
-                string precedingNamespace = precedingNamespaceSymbol.ToDisplayString(SymbolDisplayFormat.CSharpErrorMessageFormat);
+        private static void ProcessUsings(SyntaxNodeAnalysisContext context, SyntaxList<UsingDirectiveSyntax> usings)
+        {
+            var usingsDirectivesWithoutAliasAndStatic = usings.Where(ExcludeAliasAndStaticUsingDirectives);
+            var systemUsingDirectives = usingsDirectivesWithoutAliasAndStatic.Where(directive => directive.IsSystemUsingDirective());
+            var usingDirectives = usingsDirectivesWithoutAliasAndStatic.Where(directive => !directive.IsSystemUsingDirective());
```

Comment:
:bulb: This is not really efficient, as it is doing multiple iterations over the same (part of the) collection.

I would suggest something like:
```C#
var usingDirectives = new List<UsingDirectiveSyntax>();
var systemUsingDirectives = new List<UsingDirectiveSyntax>();

foreach (var usingDirective in usings)
{
    if (IsAliasorStaticUsingDirective(usingDirective))
    {
        continue;
    }

    if (usingDirective.IsSystemUsingDirective())
    {
        systemUsingDirectives.Add(usingDirective);
    }
    else
    {
        usingDirectives,Add(usingDirective);
    }
}

CheckIncorrectlyOrderedUsingsAndReportDiagnostic(context, usingDirectives);
CheckIncorrectlyOrderedUsingsAndReportDiagnostic(context, systemUsingDirectives);

```